orchestrator/graphql/schemas/subscription.py

```diff
@@ -0,0 +1,172 @@
+from orchestrator.schemas.base import OrchestratorBaseModel
+from orchestrator.schemas.product import ProductSchema
+from orchestrator.schemas.subscription_descriptions import SubscriptionDescriptionSchema
+from orchestrator.services.subscriptions import build_extended_domain_model
+from orchestrator.utils.helpers import to_camel
+
+
+@strawberry.type
+class SubscriptionProductBlock:
+    id: int
+    parent: Optional[int]
+    owner_subscription_id: UUID
+    resource_types: JSON
+
+
+def is_product_block(candidate: Any) -> bool:
+    if isinstance(candidate, dict):
+        # TODO: also filter on tag (needs addition of tag in orchestrator endpoint)
+        # NOTE: this crosses subscription boundaries. If needed we can add an additional filter to limit that.
+        return candidate.get("owner_subscription_id", None)
+    return False
+
+
+def get_all_product_blocks(subscription: dict[str, Any], _tags: Optional[list[str]]) -> list[dict[str, Any]]:
+    gen_id = count()
+
+    def locate_product_block(candidate: dict[str, Any]) -> Generator:
+        def new_product_block(item: dict[str, Any]) -> Generator:
+            enriched_item = item | {"id": next(gen_id), "parent": candidate.get("id")}
+            yield enriched_item
+            yield from locate_product_block(enriched_item)
+
+        for value in candidate.values():
+            if is_product_block(value):
+                yield from new_product_block(value)
+            elif isinstance(value, list):
+                for item in value:
+                    if is_product_block(item):
+                        yield from new_product_block(item)
+
+    return list(locate_product_block(subscription))
+
+
+async def get_subscription_product_blocks(
+    subscription_id: UUID, tags: Optional[list[str]] = None, resource_types: Optional[list[str]] = None
+) -> list[SubscriptionProductBlock]:
+    subscription_model = SubscriptionModel.from_subscription(subscription_id)
+    subscription = build_extended_domain_model(subscription_model)
+
+    def to_product_block(product_block: dict[str, Any]) -> SubscriptionProductBlock:
+        def is_resource_type(candidate: Any) -> bool:
+            return isinstance(candidate, (bool, str, int, float, type(None)))
+
+        def requested_resource_type(key: str) -> bool:
+            return not resource_types or key in resource_types
+
+        def included(key: str, value: Any) -> bool:
+            return is_resource_type(value) and requested_resource_type(key) and key not in ("id", "parent")
+
+        return SubscriptionProductBlock(
+            id=product_block["id"],
+            parent=product_block.get("parent"),
+            owner_subscription_id=product_block["owner_subscription_id"],
+            resource_types={to_camel(k): v for k, v in product_block.items() if included(k, v)},
+        )
+
+    product_blocks = (to_product_block(product_block) for product_block in get_all_product_blocks(subscription, tags))
```

Comment:
Edit: I see that we're just returning a list of product blocks, so it will be even easier

Something like;

```
for path in product_block_paths(subscription):
    product_block_dict = getattr_in(subscription, path)
    to_product_block(product_block_dict)
```

Reply:
not just the product blocks, it does have a aditional id and parent ids